docs/nia: overview of config streamlining deprecations (#12193)
hashicorp/consul: website/content/docs/nia/release-notes/0-5-0.mdx

---
layout: docs
page_title: CTS v0.5.0
description: >-

>   Consul-Terraform-Sync Configuration Deprecations for v0.5.0

Reviewer: ```suggestion
  Consul-Terraform-Sync release notes for v0.5.0
```
I wonder if we should keep this more generic. Or is the intention for these pages to always be for deprecations?

Author: Good question. I'm not sure what the future of these pages will be! Right now, I was thinking it may be any miscellaneous release-specific documentation.

One hesitation I have about changing the title to "release notes" is that it only covers a subset of changes in the release. Maybe, naming the section "Release Notes" isn't a good idea. Let me know if you have thoughts

Reviewer: I imagine in the future there might be other changes in these pages besides deprecation. In the future, others will look at this page as a starting point, so whatever we decided will be the pattern in the future.

Author: Makes sense. What are your thoughts on going with your suggestion of "Consul-Terraform-Sync Release Notes for v0.5.0" and then below add a sentence to clarify, e.g. "The release notes for v0.5.0 will focus on a subset of configuration deprecations."

---

# Configuration Deprecations in v0.5

> We introduced a number of configuration deprecations in Consul-Terraform-Sync (CTS) v0.5. Many of these deprecations are part of an effort to streamline CTS task configuration and reduce the inconsistencies. As CTS supported more use-cases, the inconsistencies grew and required a lot of documentation to explain.

Reviewer: ```suggestion
We introduced a number of configuration deprecations in Consul-Terraform-Sync (CTS) v0.5. Many of these deprecations are part of an effort to streamline CTS task configuration and reduce the inconsistencies. As CTS support for more use-cases increased,  inconsistencies arised and required a lot of documentation to explain.
```
Minor language cleanup.

> This documentation outlines these deprecations and examples on how to upgrade your configuration for v0.5. The deprecated configurations will not be removed until v0.7 and later. We plan to also release a tool to automate upgrading your configuration to hopefully make the upgrade quicker and smoother.

Reviewer: ```suggestion
This documentation outlines the deprecations and examples on how to upgrade your configuration for v0.5. The deprecated configurations will not be removed until v0.7 and later. We plan to also release a tool to automate upgrading your configurations to make the upgrade quicker and smoother.
```
I removed hope. We want to sound confident.  We know the tool will help because you are all awesome 😄

## Overview of Deprecations

> The deprecations attempt to address two major inconsistencies. First is the use of the terminology "source" in various configuration blocks and fields. The second is the exception to configuring a task condition or module input using the `services` field and related its `service` block

Reviewer: ```suggestion
The deprecations address two major inconsistencies. First is the use of the terminology "source" in various configuration blocks and fields. The second is the exception to configuring, a task condition, or module input using the `services` field and related its `service` block
```

Removed the word "attempt". Again, let's be confident in our writing 😄

Author: I'm realizing I have some typos in "The second..." sentence which are confusing. Let me know if you have thoughts on:

> The second is the exception to configuring a task condition or a task module input by using the `services` field and its related `service` block..

Reviewer: That makes more sense 👍🏼

### Address "source" terminology

> The term "source" is used frequently in the task configuration. "Source" commonly refers to the Terraform module though it is not immediately clear and requires additional documentation to explain.

Reviewer: ```suggestion
The term "source" is used frequently in the task configuration. "Source" commonly refers to the Terraform module though it is not immediately clear and requires additional documentation to explain.
```

Should all of these sources be `source` instead? They are a configuration parameter.

Author: They are supposed to be referring to the word "source" in various configuration parameters i.e. `source`, `source_input`, and `source_includes_var`.

Maybe:

> The word "source" is used in the name of a number of task configuration. "Source" in the configuration names commonly refer to...

| Deprecated Config | Replacement Config | Description |
| --- | --- | --- |
| `source` field | `module` field | The path to the Terraform module |
| `source_input` block | `module_input` block | The configuration block for additional module input for the Terraform module |
| `source_includes_var` field | `use_as_module_input` field | The field on a `condition` block to indicate whether the condition object should be used as module input |

Removal date:
- `source`: future major release after v0.7
- `source_input`: v0.7
- `source_includes_var`: v0.7

The documentation below will refer to the deprecated configurations by their new name.

### Address `services` field and `service` block

The `services` field and its associated `service` block are frequently an exception to task configuration use-cases and add complexity and restrictions.

The `services` field is the only task condition that is not configured through the `condition` block. It is also the only additional module input that is not configured through a `module_input` block. This leads to a lot of inconsistent relationships with other `condition` and `module_input` blocks.

> We deprecated `services` field and its associated `service` filter to remove these exceptions and complicated relationship with other `condition` and `module_input` blocks:

Reviewer: ```suggestion
We deprecated the `services` field and its associated `service` filter to remove these exceptions and complicated relationship with other `condition` and `module_input` blocks:
```

## Further Details on Deprecations

### Deprecate `source` field

The `source` field in the task configuration is deprecated and can be directly renamed to `module`.

Example:

```diff
task {
  name     = "services_task"
  services = ["api", "web"]
- source   = "path/to/module"
+ module   = "path/to/module"
}
```

### Deprecate `source_input` block

The `source_input` block in the task configuration is deprecated and can be directly renamed to `module_input`.

Example:

```diff
task {
  name   = "scheduled_task"
  module = "path/to/module"
  condition "schedule" {
    cron = "* * * * Mon"
  }
- source_input "consul-kv" {
+ module_input "consul-kv" {
    path = "my_key"
  }
}
```

### Deprecate `source_includes_var` field

The `condition` block's `source_includes_var` field is deprecated for all types of conditions and can be directly renamed to `use_as_module_input`.

Example:

```diff
task {
  name   = "catalog_services_task"
  module = "path/to/module"
  condition "catalog-services" {
    regexp              = "api"
-   source_includes_var = true
+   use_as_module_input = true
  }
}

> ```

Reviewer: Same feedback as above.

+ }
}
```

**Deprecate `services` field as module input only**

When the `services` field only provides module input, it can be replaced with a `module_input "services"` block with the `names` field set to the list of the services.

Example:

```diff
task {
  name   = "services_module_input_task"
  module = "path/to/module"
  condition "consul-kv" {
    path = "my_key"
  }
- services = ["api", "web"]
+ module_input "services" {
+   names = ["api", "web"]
+ }
}
```

### Deprecate service block

The `service` block is a configuration block outside of the task block. The fields of the `service` block provides additional filtering to any task that has that service set in the `services` field.

Once the `services` field is upgraded to a `condition` or `module_input` block, as described in the section above, the deprecated `service` block's fields can be moved to the new `condition` or `module_input` block.

Example:

```diff
-services {
- name       = "api"
- datacenter = "dc2"
- namespace  = "ns"
-}

task {
  name   = "services_filter_task"
  module = "path/to/module"
  condition "services" {
    names = ["api"]
+   datacenter = "dc2"
+   namespace  = "ns"
  }
}

> ```

Author: Thanks for letting me know! I'll take a look and try reformatting with CodeTabs

> ```

Reviewer: @lornasong slack me if you need help as these components can be confusing to use at first.

This documentation outlines the deprecations and examples on how to upgrade your configuration for v0.5. The deprecated configurations will not be removed until v0.7 and later. We plan to also release a tool to automate upgrading your configurations to make the upgrade quicker and smoother.

## Overview of Deprecations

The deprecations address two major inconsistencies. First is the use of the terminology "source" in various configuration blocks and fields. The second is the exception to configuring a task condition or a task module input by using the `services` field and its related `service` block.

### Address "source" terminology

The word "source" is used in the name of a number of task configurations. "Source" in the configuration name commonly refers to the Terraform module though it is not immediately clear and requires additional documentation to explain.

We deprecated all configuration with the term "source" and replaced it with "module" in the task configuration:

| Deprecated Config | Replacement Config | Description |
| --- | --- | --- |
| `source` field | `module` field | The path to the Terraform module |

> | `source_input` block | `module_input` block | The configuration block for additional module input for the Terraform module |

Reviewer 2: Would it be worth calling out here that additional module inputs are Terraform input variables?

Author: Sure! I tried rewording it as below to capture your thoughts. Let me know any feedback!

> The configuration block to define the values for additional Terraform input variables for the Terraform module

> | `source_includes_var` field | `use_as_module_input` field | The field on a `condition` block to indicate whether the condition object should be used as module input |

Reviewer 2: ```suggestion
| `source_includes_var` field | `use_as_module_input` field | The field on a `condition` block to indicate whether the same values of the condition object should be used for the module input |
```
I was thinking it would be more clear if we say that it's the same values rather than the same object, but then we lose the "used as module input" in the description. 🤔  So feel free to disregard if you prefer what you have!

Author: I see what you're saying - it's inaccurate to say that its the same object. Let me know your thoughts on the below. Not sure if the "as" doesn't fit in place of the "for"

> The field on the `condition` block to indicate whether the values of the condition object should also be used as module input

Reviewer 2: I like your wording here!

| `service` block | fields moved into the `condition "services"` and `module_input "services"` block | The configuration block to provide additional filtering on the services in the `services` field |

The `services` field and its associated `service` block will be removed in a future major release after v0.7.

## Further Details on Deprecations

### Deprecate `source` field

The `source` field in the task configuration is deprecated and can be directly renamed to `module`.

Example:

<CodeTabs heading="Deprecate `source` field" tabs={["Deprecated", "New"]}>

<CodeBlockConfig highlight="4">

```hcl
task {
  name     = "services_task"
  services = ["api", "web"]
  source   = "path/to/module"
}
```

</CodeBlockConfig>

<CodeBlockConfig highlight="4">

```hcl
task {
  name     = "services_task"
  services = ["api", "web"]
  module   = "path/to/module"
}
```

</CodeBlockConfig>

</CodeTabs>

### Deprecate `source_input` block

The `source_input` block in the task configuration is deprecated and can be directly renamed to `module_input`.

Example:

> <CodeTabs heading="Deprecate `source_input` block" tabs={["Deprecated", "New"]}>

Reviewer: ```suggestion
<CodeTabs heading="Deprecate source_input block" tabs={["Deprecated", "New"]}>
```

<CodeBlockConfig highlight="7">

```hcl
task {
  name   = "scheduled_task"
  module = "path/to/module"
  condition "schedule" {
    cron = "* * * * Mon"
  }
  source_input "consul-kv" {
    path = "my_key"
  }
}
```

</CodeBlockConfig>

<CodeBlockConfig highlight="7">

```hcl
task {
  name   = "scheduled_task"
  module = "path/to/module"
  condition "schedule" {
    cron = "* * * * Mon"
  }
  module_input "consul-kv" {
    path = "my_key"
  }
}
```

</CodeBlockConfig>

</CodeTabs>

### Deprecate `source_includes_var` field

> The `condition` block's `source_includes_var` field is deprecated for all types of conditions and can be directly renamed to `use_as_module_input`.

Reviewer: ```suggestion
The `condition` block's source_includes_var field is deprecated for all types of conditions and can be directly renamed to `use_as_module_input`.
```

Author: Wanted to double-check removing the backticks on here? The "source_includes_var" here refers to the configuration. It's placed in text rather than the header CodeTabs header. Let me know if I'm missing something

Reviewer: @lornasong  just in the headings for the codetabs `heading="blah blah `source.....`"

Author: Yes, this suggestion is actually not in the header of CodeTabs but is in a text block. It looks like it might be unintentional and wanted to double-check.

Author: No worries! Wanted to make sure I didn't misunderstand. Thanks for checking.

Example:

> <CodeTabs heading="Deprecate `source_includes_var` field" tabs={["Deprecated", "New"]}>

Reviewer: ```suggestion
<CodeTabs heading="Deprecate source_includes_var field" tabs={["Deprecated", "New"]}>
```

<CodeBlockConfig highlight="6">

```hcl
task {
  name   = "catalog_services_task"
  module = "path/to/module"
  condition "catalog-services" {
    regexp              = "api"
    source_includes_var = true
  }
}
```

</CodeBlockConfig>

<CodeBlockConfig highlight="6">

```hcl
task {
  name   = "catalog_services_task"
  module = "path/to/module"
  condition "catalog-services" {
    regexp              = "api"
    use_as_module_input = true
  }
}
```

</CodeBlockConfig>

</CodeTabs>

### Deprecate `services` field

The `services` field can play different roles in a task depending on other task configurations:
- When no `condition` block is configured, then the `services` field is the task's condition and module input
- When there is a `condition` block configured, then the `services field is only module input for the task

The `services` field can be replaced depending on the role it has for the task.

**Deprecate `services` field as a condition**

When the `services` field behaves as a condition, it can be replaced with a `condition "services"` block with the `names` field set to the list of the services.

Example:

> <CodeTabs heading="Deprecate `services` field as a condition" tabs={["Deprecated", "New"]}>

Reviewer: ```suggestion
<CodeTabs heading="Deprecate services field as a condition" tabs={["Deprecated", "New"]}>
```

<CodeBlockConfig highlight="4">

```hcl
task {
  name     = "services_condition_task"
  module   = "path/to/module"
  services = ["api", "web"]
}
```

</CodeBlockConfig>

<CodeBlockConfig highlight="4,5,6">

```hcl
task {
  name     = "services_condition_task"
  module   = "path/to/module"
  condition "services" {
    names = ["api", "web"]

>   }

Reviewer 2: Can we call out in this section that this task is valid because `use_as_module_input` is set to true by default? It may also help to provide an example where we're explicitly setting it to `true`.

Author: I realized for `condition "services"` that `use_as_module_input` field was introduced in 0.5. I couldn't get the right wording for a callout here.

I ended up adding `use_as_module_input` in the example below for `condition "consul-kv"` block and added a little callout there. Let me know your thoughts!

Note: apologies for all push noise - ran into some issues with the preview deploy and caught some typos!

Reviewer 2: The callout looks good to me!

Example:
<CodeTabs heading="Deprecate `services` field as a module input" tabs={["Deprecated", "New"]}>

<CodeBlockConfig highlight="7">

```hcl
task {
  name   = "services_module_input_task"
  module = "path/to/module"
  condition "consul-kv" {
    path = "my_key"
  }
  services = ["api", "web"]
}
```

</CodeBlockConfig>

<CodeBlockConfig highlight="7,8,9">

```hcl
task {
  name   = "services_module_input_task"
  module = "path/to/module"
  condition "consul-kv" {
    path = "my_key"
  }
  module_input "services" {
    names = ["api", "web"]
  }
}
```

</CodeBlockConfig>

</CodeTabs>

### Deprecate service block

The `service` block is a configuration block outside of the task block. The fields of the `service` block provides additional filtering to any task that has that service set in the `services` field.

Once the `services` field is upgraded to a `condition` or `module_input` block, as described in the section above, the deprecated `service` block's fields can be moved to the new `condition` or `module_input` block.

Example:

> <CodeTabs heading="Deprecate `service` block" tabs={["Deprecated", "New"]}>

Reviewer: ```suggestion
<CodeTabs heading="Deprecate service block" tabs={["Deprecated", "New"]}>
```

<CodeBlockConfig highlight="9,10,11,12,13">

```hcl
task {
  name   = "services_filter_task"
  module = "path/to/module"
  condition "services" {
    names = ["api"]
  }
}

services {
 name       = "api"
 datacenter = "dc2"
 namespace  = "ns"
}
```

</CodeBlockConfig>

<CodeBlockConfig highlight="6,7">

```hcl
task {
  name   = "services_filter_task"
  module = "path/to/module"
  condition "services" {
    names = ["api"]
    datacenter = "dc2"
    namespace  = "ns"
  }
}
```

</CodeBlockConfig>

</CodeTabs>

There is an edge-case that requires a more complicated deprecation upgrade. Some tasks may have multiple services in the `services` field and a `service` block configured for each of those services. When the fields are different across the `service` blocks, it is necessary to split the task and create new, separate tasks for each service.

Example:

> <CodeTabs heading="Deprecate `service` block edge-case" tabs={["Deprecated", "New"]}>

Reviewer: ```suggestion
<CodeTabs heading="Deprecate service block edge-case" tabs={["Deprecated", "New"]}>
```
